frontend/src/Frontend/UI/Dialogs/Send.hs

```diff
@@ -1026,49 +1055,62 @@ listenForContinuation
   -> [S.ClientEnv]
   -> Event t Pact.RequestKey
   -> m (Event t (Either Text (Pact.RequestKey, Pact.PactExec)))
-listenForContinuation model envs requestKey = performEventAsync $ ffor requestKey $ \rk cb -> liftJSM $ forkJSM $ do
-  r <- doReqFailover envs (Api.listen Api.apiV1Client $ Api.ListenerRequest rk) >>= \case
-    Left es -> packHttpErrors (model ^. logger) es
-    Right (Api.ListenResponse cr) -> case Pact._crContinuation cr of
-      Nothing ->
-        pure $ Left $ T.unlines ["Result was not a continuation", tshow (Pact._crResult cr)]
-      Just pe ->
-        pure $ Right (Pact._crReqKey cr, pe)
-
-    Right (Api.ListenTimeout _) -> pure $ Left "Listen timeout"
-  liftIO $ cb r
-
-packHttpErrors :: MonadIO m => Logger t -> [S.ClientError] -> m (Either Text a)
-packHttpErrors logL es = do
-  let prettyErr = T.unlines $ fmap (tshow . packHttpErr) es
-  putLog logL LevelWarn prettyErr
-  pure $ Left prettyErr
-
--- | Listen to a request key for the second step (redeeming coin on the target
--- chain).
-listenForSuccess
+pollNodesForCont model envs reqEv = pollNodesForReq model envs ((Nothing,) <$> reqEv) 12 15 $ \cr _->
+  case Pact._crContinuation cr of
+    Nothing ->
+      Left $ T.unlines ["Result was not a continuation", tshow (Pact._crResult cr)]
+    Just pe ->
+      Right (Pact._crReqKey cr, pe)
+
+data PollingAttempt a =
+    PollingAttempt_Error Text
+  | PollingAttempt_NotFound
+  | PollingAttempt_Success a
+  deriving (Eq)
+
+pollNodesForReq
   :: ( TriggerEvent t m
      , PerformEvent t m
      , MonadJSM (Performable m)
      , HasLogger model t
      )
   => model
   -> [S.ClientEnv]
-  -> Event t (Pact.PactId, Pact.RequestKey)
-  -> m (Event t (Either Text ()))
-listenForSuccess model envs e = performEventAsync $ ffor e $ \(PactId pactId, requestKey) cb -> liftJSM $ forkJSM $ do
-  -- This string appears in the error if the pact was previously completed
-  let completedMsg = "resumePact: pact completed: " <> pactId
-  r <- doReqFailover envs (Api.listen Api.apiV1Client $ Api.ListenerRequest requestKey) >>= \case
-    Left es -> packHttpErrors (model ^. logger) es
-    Right (Api.ListenResponse cr) -> case Pact._crResult cr of
-      Pact.PactResult (Left pe)
-        -- There doesn't seem to be a nicer way to do this check
-        | completedMsg == tshow (peDoc pe) -> pure $ Right ()
-        | otherwise -> pure $ Left $ tshow $ peDoc pe
-      Pact.PactResult (Right _) -> pure $ Right ()
-    Right (Api.ListenTimeout _) -> pure $ Left "Listen timeout"
-  liftIO $ cb r
+  -> Event t (Maybe Pact.PactId, Pact.RequestKey)
+  -> Int --max polls
+  -> Int -- delay between polls (seconds)
+  -> (Pact.CommandResult Hash -> Maybe Pact.PactId -> Either Text a) -- callback for handling poll result
+  -> m (Event t (Either Text a))
+pollNodesForReq model envs requestPair maxPolls waitTime handler =
```

Comment:
If you don't mind me suggesting, have you used the `retry` library before? It is really good for forming delay combinators for such situations such as these?

Reply:
Yeah, I like the `retry` library as well but in this case I'm not sure it's worth it.  We don't need sophisticated delay strategies here and it's since it's a web app I'd probably say that incurring another dependency is worse than the relatively straightforward loop to implement this logic.